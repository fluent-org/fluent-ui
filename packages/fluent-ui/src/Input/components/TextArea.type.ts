import { StandardProps } from '../..'
import * as React from 'react'
import * as PropTypes from 'prop-types'

export type TextAreaClassProps = 'root' | 'error' | 'ghost' | 'resize'

type Resize = 'none' | 'both' | 'horizontal' | 'vertical' | 'block' | 'inline'

export interface TextAreaProps
  extends StandardProps<
    React.TextareaHTMLAttributes<HTMLTextAreaElement>,
    HTMLTextAreaElement,
    'onChange'
  > {
  value?: string
  onChange?: (value: string) => void
  placeholder?: string
  disabled?: boolean
  error?: boolean
  ghost?: boolean
  resize?: Resize
}

export const TextAreaPropTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  ghost: PropTypes.bool,
  resize: PropTypes.oneOf<Resize>(['none', 'both', 'horizontal', 'vertical', 'block', 'inline'])
}
