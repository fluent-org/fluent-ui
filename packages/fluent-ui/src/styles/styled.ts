import styled from 'styled-components'
import { th, variant, system, breakpoints } from '@xstyled/system'

export { styled, system, th, variant, breakpoints }

export * from 'styled-components'
export * from '@xstyled/system'
// export * from '@xstyled/prop-types'

export * from './stylize'
