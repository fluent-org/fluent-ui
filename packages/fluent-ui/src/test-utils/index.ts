export * from '@testing-library/react'
// @ts-ignore
export * from './createRender'
