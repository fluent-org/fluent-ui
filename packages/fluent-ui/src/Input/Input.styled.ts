import { Style, Styles } from 'jss'
import { Theme } from '../styles'
import { InputClassProps, InputProps } from './Input.type'

const PADDING_RIGHT = 34

const wrapper = (theme: Theme): Style => (
  props: InputProps & { focus: boolean; _value: string }
): Style => ({
  display: 'inline-block',
  position: 'relative',
  minWidth: 180,
  '& > span': {
    fontWeight: 600,
    position: 'absolute',
    left: 0,
    top: 0,
    cursor: 'text',
    transition: theme.transitions!.inputLabel,
    transformOrigin: 'top left',
    ...theme.sizes!.medium!.inputLabel,
    color: props.focus ? theme.colors!.black!.default : theme.colors!.standard!.dark2,
    ...((props.focus || props._value) && theme.sizes!.medium!.inputLabelFocus)
  },
  '& > div': { margin: '0 !important' },
  ...(!props.disabled && {
    '&:hover': {
      '& input': {
        borderColor: theme.colors!.standard!.transparent1
      }
    }
  })
})
const disabled: Style = {
  cursor: 'not-allowed'
}

const root = (theme: Theme): Style => (
  props: InputProps & { focus: boolean; _value: string }
): Style => ({
  boxSizing: 'border-box',
  outline: 'none',
  font: 'inherit',
  width: '100%',
  borderRadius: 2,
  border: '2px solid',
  borderColor: 'transparent',
  transition: theme.transitions!.input,
  backgroundColor: theme.colors!.standard!.light2,
  '&::placeholder': {
    transition: theme.transitions!.input,
    ...(!props.disabled && {
      color: 'transparent'
    })
  },
  '&:active, &:focus': {
    borderColor: `${props.error ? '' : theme.colors!.black!.default} ${
      props.disabled ? '' : '!important'
    }`,
    backgroundColor: theme.colors!.white!.default,
    ...(!props.disabled && {
      '&::placeholder': {
        color: theme.colors!.standard!.dark2
      }
    })
  },
  '&:disabled': {
    color: theme.colors!.standard!.dark3,
    backgroundColor: theme.colors!.standard!.default,
    borderColor: theme.colors!.standard!.default,
    pointerEvents: 'none'
  },
  ...(props.label ? theme.sizes!.medium!.inputWithLabel : theme.sizes!.medium!.input),
  '&:not(:last-child)': {
    paddingRight: PADDING_RIGHT
  }
})
const error = (theme: Theme): Style => ({
  borderColor: `${theme.colors!.error!.default} !important`,
  '&:hover': {
    borderColor: `${theme.colors!.error!.default} !important`
  },
  '&:active, &:focus': {
    borderColor: `${theme.colors!.error!.default} !important`
  }
})

const extra = (theme: Theme): Style => ({
  position: 'absolute',
  top: '50%',
  transform: 'translateY(-50%)',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  color: theme.colors!.standard!.transparent3
})
const clearedIcon = (theme: Theme): Style => ({
  ...extra(theme),
  right: 6
})
const suffix = (theme: Theme): Style => ({
  ...extra(theme),
  right: 6
})

const ghost = (theme: Theme): Style => ({
  backgroundColor: `${theme.colors!.standard!.transparent1} !important`,
  borderColor: theme.colors!.standard!.transparent1,
  '&:hover': {
    borderColor: theme.colors!.standard!.transparent2
  },
  '&:active, &:focus': {
    borderColor: theme.colors!.standard!.transparent3
  }
})

export const styles = (theme: Theme): Styles<InputClassProps> => ({
  wrapper: wrapper(theme),
  disabled,
  root: root(theme),
  error: error(theme),
  clearedIcon: clearedIcon(theme),
  suffix: suffix(theme),
  ghost: ghost(theme)
})
