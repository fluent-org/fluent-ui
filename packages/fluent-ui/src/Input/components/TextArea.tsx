import * as React from 'react'
import classNames from 'classnames'
import { createUseStyles } from '@fluent-ui/styles'
import { styles } from './TextArea.styled'
import { Theme } from '../../styles'
import { TextAreaClassProps, TextAreaProps, TextAreaPropTypes } from './TextArea.type'

export const name = 'TextArea'

const useStyles = createUseStyles<Theme, TextAreaClassProps>(styles, { name })

const TextArea: React.FC<TextAreaProps> = React.forwardRef<HTMLTextAreaElement, TextAreaProps>(
  (props, ref): React.ReactElement => {
    const {
      className: classNameProp,
      value,
      onChange,
      placeholder,
      disabled,
      error,
      ghost,
      resize = 'both',
      ...rest
    } = props

    const handleChange = React.useCallback(
      (e: React.ChangeEvent<HTMLTextAreaElement> | null): void => {
        onChange && onChange(e ? e.target.value : '')
      },
      [onChange]
    )

    const classes = useStyles(props)
    const className = classNames(
      classes.root,
      {
        [classes.error]: error,
        [classes.ghost]: ghost,
        [classes.resize]: !!resize
      },
      classNameProp
    )

    return (
      <textarea
        className={className}
        ref={ref}
        value={value}
        onChange={handleChange}
        placeholder={placeholder}
        disabled={disabled}
        {...rest}
      />
    )
  }
)

TextArea.displayName = `F${name}`

TextArea.propTypes = TextAreaPropTypes

TextArea.defaultProps = {
  resize: 'both'
}

export default TextArea
