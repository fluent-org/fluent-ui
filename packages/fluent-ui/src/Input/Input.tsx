import * as React from 'react'
import classNames from 'classnames'
import { createUseStyles } from '@fluent-ui/styles'
import { styles } from './Input.styled'
import { Theme } from '../styles'
import { InputClassProps, InputProps, InputPropTypes, InputType } from './Input.type'
import { toUpperCase } from '../utils'

import FormLabel from '../FormLabel'
import { CloseLine as CloseLineIconIcon } from '@fluent-ui/icons'
import Transition from '../Transition'
import IconButton from '../IconButton'
import TextArea from './components/TextArea'

export const name = 'Input'

const useStyles = createUseStyles<Theme, InputClassProps>(styles, { name })

const Input: React.FC<InputProps> = React.forwardRef<HTMLInputElement, InputProps>(
  (props, ref): React.ReactElement => {
    const {
      className: classNameProp,
      style,
      value,
      onChange,
      onFocus,
      onBlur,
      label,
      placeholder,
      disabled,
      cleared,
      password,
      error,
      // TODO cleared 和 suffix 同时存在时应该 warning
      suffix = null,
      ghost,
      ...rest
    } = props

    const clearedRef = React.useRef<HTMLButtonElement>(null)
    const [_value, _setValue] = React.useState(value)
    const handleChange = React.useCallback(
      (e: React.ChangeEvent<HTMLInputElement> | null): void => {
        onChange && onChange(e ? e.target.value : '')
        _setValue(e ? e.target.value : '')
      },
      [onChange]
    )
    const handleClear = React.useCallback(
      (e: React.MouseEvent<HTMLElement>): void => {
        e.preventDefault()
        e.stopPropagation()
        handleChange(null)
        // TODO 应该引入 warning 当前组件为非受控组件时提示
      },
      [handleChange]
    )

    const _label = React.useMemo((): string => (label ? toUpperCase(label) : ''), [label])
    const [focus, setFocus] = React.useState(false)
    const handleFocus = React.useCallback(
      (e): void => {
        typeof onFocus === 'function' && onFocus(e)
        setFocus(true)
      },
      [onFocus]
    )
    const handleBlur = React.useCallback(
      (e): void => {
        typeof onBlur === 'function' && onBlur(e)
        setFocus(false)
      },
      [onBlur]
    )

    const classes = useStyles({ ...props, focus, _value })
    const wrapperClassName = classNames(classes.wrapper, {
      [classes.disabled]: disabled
    })
    const className = classNames(
      classes.root,
      {
        [classes.error]: error,
        [classes.ghost]: ghost
      },
      classNameProp
    )

    return (
      <FormLabel className={wrapperClassName} label={_label} style={style}>
        <div>
          <input
            className={className}
            ref={ref}
            value={value}
            onChange={handleChange}
            onFocus={handleFocus}
            onBlur={handleBlur}
            placeholder={placeholder}
            disabled={disabled}
            type={password ? 'password' : 'text'}
            {...rest}
          />
          {suffix && <div className={classes.suffix}>{suffix}</div>}
          {cleared && (
            <Transition visible={!!value} wrapper={false}>
              <IconButton
                className={classes.clearedIcon}
                onClick={handleClear}
                size="small"
                ref={clearedRef}
              >
                <CloseLineIconIcon />
              </IconButton>
            </Transition>
          )}
        </div>
      </FormLabel>
    )
  }
)

Object.defineProperty(Input, 'TextArea', {
  get(): typeof TextArea {
    return TextArea
  }
})

Input.displayName = `F${name}`

Input.propTypes = InputPropTypes

Input.defaultProps = {
  suffix: null
}

export default Input as InputType
