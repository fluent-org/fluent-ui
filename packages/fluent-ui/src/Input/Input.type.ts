import * as React from 'react'
import * as PropTypes from 'prop-types'
import { StandardProps } from '..'
import TextArea from './components/TextArea'

export type InputClassProps =
  | 'root'
  | 'error'
  | 'wrapper'
  | 'clearedIcon'
  | 'prefix'
  | 'suffix'
  | 'ghost'

export interface InputProps
  extends StandardProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement,
    'onChange' | 'prefix'
  > {
  value?: string
  onChange?: (value: string) => void
  placeholder?: string
  disabled?: boolean
  cleared?: boolean
  password?: boolean
  error?: boolean
  prefix?: React.ReactNode
  suffix?: React.ReactNode
  ghost?: boolean
}

export const InputPropTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  cleared: PropTypes.bool,
  password: PropTypes.bool,
  error: PropTypes.bool,
  prefix: PropTypes.node,
  suffix: PropTypes.node,
  ghost: PropTypes.bool
}

export interface InputType extends React.ForwardRefExoticComponent<InputProps> {
  TextArea: typeof TextArea
}
