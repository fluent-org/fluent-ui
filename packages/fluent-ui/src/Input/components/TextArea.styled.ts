import { Style, Styles } from 'jss'
import { Theme } from '../../styles'
import { TextAreaClassProps, TextAreaProps } from './TextArea.type'

const root = (theme: Theme): Style => ({
  outline: 'none',
  font: 'inherit',
  borderRadius: 2,
  border: '2px solid',
  borderColor: theme.colors!.standard!.default,
  transition: theme.transitions!.input,
  '&:hover': {
    borderColor: theme.colors!.standard!.dark1
  },
  '&:active, &:focus': {
    borderColor: theme.colors!.primary!.default
  },
  '&:disabled': {
    color: theme.colors!.standard!.dark2,
    backgroundColor: theme.colors!.standard!.light1,
    cursor: 'not-allowed',
    pointerEvents: 'none'
  },
  ...theme.sizes!.medium!.input
})

const error = (theme: Theme): Style => ({
  borderColor: theme.colors!.error!.default,
  '&:hover': {
    borderColor: theme.colors!.error!.default
  },
  '&:active, &:focus': {
    borderColor: theme.colors!.error!.default
  }
})

const ghost = (theme: Theme): Style => ({
  backgroundColor: 'transparent',
  borderColor: theme.colors!.standard!.transparent1,
  '&:hover': {
    borderColor: theme.colors!.standard!.transparent2
  },
  '&:active, &:focus': {
    borderColor: theme.colors!.standard!.transparent3
  }
})

const resize = ({ resize }: TextAreaProps): Style => ({
  resize
})

export const styles = (theme: Theme): Styles<TextAreaClassProps> => ({
  root: root(theme),
  error: error(theme),
  ghost: ghost(theme),
  resize
})
