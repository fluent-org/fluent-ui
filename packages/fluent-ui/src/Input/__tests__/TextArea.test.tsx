import * as React from 'react'
import { fireEvent } from '@testing-library/react'
import { getClasses, render } from '../../test-utils'
import '@testing-library/jest-dom/extend-expect'

import Input from '../Input'
import { name } from '../components/TextArea'
import { styles } from '../components/TextArea.styled'
import { TextAreaClassProps } from '../components/TextArea.type'

const classes = getClasses<TextAreaClassProps>(styles, name)

describe('Input.TextArea', (): void => {
  const text = 'Hint Text'

  test('should be support ref', (): void => {
    const ref = React.createRef<HTMLTextAreaElement>()
    const { getByTestId } = render(<Input.TextArea ref={ref} data-testid="textarea" />)
    const { current: element } = ref
    expect(element).toEqual(getByTestId('textarea'))
  })

  test('should be controlled', (): void => {
    const ref = React.createRef<HTMLTextAreaElement>()
    let value = ''
    function handleChange(v: string): void {
      value = v
    }
    render(<Input.TextArea ref={ref} value={value} onChange={handleChange} />)
    fireEvent.change(ref.current!, { target: { value: text } })
    expect(value).toEqual(text)
    expect(ref.current!).toHaveClass(classes.root)
  })

  test('should be support placeholder', (): void => {
    const ref = React.createRef<HTMLTextAreaElement>()
    render(<Input.TextArea ref={ref} placeholder={text} />)
    expect((ref.current as HTMLTextAreaElement).placeholder).toEqual(text)
  })

  test('should be support disabled', (): void => {
    const ref = React.createRef<HTMLTextAreaElement>()
    render(<Input.TextArea ref={ref} disabled={true} />)
    expect((ref.current as HTMLTextAreaElement).disabled).toEqual(true)
  })

  test('should be support error', (): void => {
    const ref = React.createRef<HTMLTextAreaElement>()
    const { sheets } = render(<Input.TextArea ref={ref} error={true} />)
    expect(ref.current as HTMLTextAreaElement).toHaveClass(classes.error)
    expect(sheets.toString()).toMatchSnapshot()
  })

  test('should be support ghost', (): void => {
    const ref = React.createRef<HTMLTextAreaElement>()
    const { sheets } = render(<Input.TextArea ref={ref} ghost={true} />)
    expect(ref.current as HTMLTextAreaElement).toHaveClass(classes.ghost)
    expect(sheets.toString()).toMatchSnapshot()
  })
})
