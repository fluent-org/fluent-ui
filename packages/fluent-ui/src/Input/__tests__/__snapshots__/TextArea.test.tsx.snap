// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.TextArea should be support error 1`] = `
".FTextArea-root {
  font: inherit;
  border: 2px solid;
  outline: none;
  padding: 11px 6px;
  transition: background-color 250ms cubic-bezier(0.7, 0.1, 0.33, 1) 0ms,color 250ms cubic-bezier(0.7, 0.1, 0.33, 1) 0ms,border-color 250ms cubic-bezier(0.7, 0.1, 0.33, 1) 0ms;
  border-color: #dedede;
  border-radius: 2px;
}
.FTextArea-root:hover {
  border-color: #b5b5b5;
}
.FTextArea-root:active, .FTextArea-root:focus {
  border-color: #0078d4;
}
.FTextArea-root:disabled {
  color: #8c8c8c;
  cursor: not-allowed;
  pointer-events: none;
  background-color: #e9e9e9;
}
.FTextArea-error {
  border-color: #f44336;
}
.FTextArea-error:hover {
  border-color: #f44336;
}
.FTextArea-error:active, .FTextArea-error:focus {
  border-color: #f44336;
}
.FTextArea-ghost {
  border-color: rgba(0, 0, 0, 0.1);
  background-color: transparent;
}
.FTextArea-ghost:hover {
  border-color: rgba(0, 0, 0, 0.2);
}
.FTextArea-ghost:active, .FTextArea-ghost:focus {
  border-color: rgba(0, 0, 0, 0.4);
}
.FTextArea-resize {}
.FTextArea-resize-0 {
  resize: both;
}"
`;

exports[`Input.TextArea should be support ghost 1`] = `
".FTextArea-root {
  font: inherit;
  border: 2px solid;
  outline: none;
  padding: 11px 6px;
  transition: background-color 250ms cubic-bezier(0.7, 0.1, 0.33, 1) 0ms,color 250ms cubic-bezier(0.7, 0.1, 0.33, 1) 0ms,border-color 250ms cubic-bezier(0.7, 0.1, 0.33, 1) 0ms;
  border-color: #dedede;
  border-radius: 2px;
}
.FTextArea-root:hover {
  border-color: #b5b5b5;
}
.FTextArea-root:active, .FTextArea-root:focus {
  border-color: #0078d4;
}
.FTextArea-root:disabled {
  color: #8c8c8c;
  cursor: not-allowed;
  pointer-events: none;
  background-color: #e9e9e9;
}
.FTextArea-error {
  border-color: #f44336;
}
.FTextArea-error:hover {
  border-color: #f44336;
}
.FTextArea-error:active, .FTextArea-error:focus {
  border-color: #f44336;
}
.FTextArea-ghost {
  border-color: rgba(0, 0, 0, 0.1);
  background-color: transparent;
}
.FTextArea-ghost:hover {
  border-color: rgba(0, 0, 0, 0.2);
}
.FTextArea-ghost:active, .FTextArea-ghost:focus {
  border-color: rgba(0, 0, 0, 0.4);
}
.FTextArea-resize {}
.FTextArea-resize-0 {
  resize: both;
}"
`;
